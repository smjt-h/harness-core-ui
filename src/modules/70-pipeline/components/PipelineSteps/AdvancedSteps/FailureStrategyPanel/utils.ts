/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import type { FormikErrors, FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'
import { Intent } from '@blueprintjs/core'

import type {
  RetryFailureActionConfig,
  IgnoreFailureActionConfig,
  ManualInterventionFailureActionConfig,
  AbortFailureActionConfig,
  StepGroupFailureActionConfig,
  MarkAsSuccessFailureActionConfig,
  StageRollbackFailureActionConfig,
  FailureStrategyConfig,
  OnFailureConfig
} from 'services/cd-ng'

export type AllActions =
  | RetryFailureActionConfig
  | IgnoreFailureActionConfig
  | ManualInterventionFailureActionConfig
  | AbortFailureActionConfig
  | StepGroupFailureActionConfig
  | MarkAsSuccessFailureActionConfig
  | StageRollbackFailureActionConfig

export interface AllFailureStrategyConfig extends FailureStrategyConfig {
  onFailure: OnFailureConfig & { action: AllActions }
}

export function hasItems<T>(data?: T[]): boolean {
  return Array.isArray(data) && data.length > 0
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function findTabWithErrors(errors?: FormikErrors<any>): number {
  const fsErrors = errors?.failureStrategies

  return Array.isArray(fsErrors)
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fsErrors.findIndex((err: any) => !isEmpty(err))
    : -1
}

export interface FormState {
  failureStrategies?: AllFailureStrategyConfig[]
}

export interface HandleChangeInStrategiesProps {
  formValues: FormState
  selectedStrategyNum: number
  setSelectedStrategyNum: (value: SetStateAction<number>) => void
  setFormikState: FormikProps<FormState>['setFormikState']
}

export function handleChangeInStrategies({
  formValues,
  selectedStrategyNum,
  setSelectedStrategyNum,
  setFormikState
}: HandleChangeInStrategiesProps): void {
  const fs = formValues.failureStrategies
  /* istanbul ignore else */
  if (Array.isArray(fs)) {
    /* istanbul ignore else */
    if (selectedStrategyNum >= fs.length) {
      // select the new last tab, if the last tab was deleted
      setSelectedStrategyNum(Math.max(0, fs.length - 1))
    }

    /* istanbul ignore else */
    if (fs.length === 0) {
      // reset errors when all the tabs are deleted
      setFormikState({ errors: {}, submitCount: 0 })
    }
  }
}

export function getTabIntent(i: number, selectedStrategyNum: number): Intent {
  return i === selectedStrategyNum ? Intent.PRIMARY : Intent.NONE
}
