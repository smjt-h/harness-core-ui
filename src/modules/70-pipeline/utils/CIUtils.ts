/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get, isEmpty } from 'lodash-es'
import moment from 'moment'
import { RUNTIME_INPUT_VALUE, SelectOption } from '@harness/uicore'
import type { GitFilterScope } from '@common/components/GitFilters/GitFilters'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { RegExAllowedInputExpression } from '@pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet'
import type { GitQueryParams } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { ConnectorRefWidth } from './constants'

// Returns first 7 letters of commit ID
export function getShortCommitId(commitId: string): string {
  return commitId.slice(0, 7)
}

// TODO: Add singular forms, better using i18n because they have support for it
export function getTimeAgo(timeStamp: number): string {
  const currentDate = moment(new Date())
  const timeStampAsDate = moment(timeStamp)

  if (currentDate.diff(timeStampAsDate, 'days') > 30) {
    return `on ${timeStampAsDate.format('MMM D')}`
  } else if (currentDate.diff(timeStampAsDate, 'days') === 1) {
    return 'yesterday'
  } else if (currentDate.diff(timeStampAsDate, 'days') === 0) {
    if (currentDate.diff(timeStampAsDate, 'minutes') >= 60) {
      return `${currentDate.diff(timeStampAsDate, 'hours')} hours ago`
    } else {
      return `${currentDate.diff(timeStampAsDate, 'minutes')} minutes ago`
    }
  } else {
    return `${currentDate.diff(timeStampAsDate, 'days')} days ago`
  }
}

export function useGitScope(): GitFilterScope | undefined {
  const gitDetails = usePipelineContext()?.state?.gitDetails
  const { repoIdentifier, branch } = useQueryParams<GitQueryParams>()

  if (!isEmpty(gitDetails)) {
    return {
      repo: gitDetails.repoIdentifier!,
      branch: gitDetails.branch!,
      getDefaultFromOtherRepo: true
    }
  } else if (!!repoIdentifier && !!branch) {
    return {
      repo: repoIdentifier,
      branch,
      getDefaultFromOtherRepo: true
    }
  }
}

export const getAllowedValuesFromTemplate = (template: Record<string, any>, fieldPath: string): SelectOption[] => {
  if (!template || !fieldPath) {
    return []
  }
  const value = get(template, fieldPath, '')
  const items: SelectOption[] = []
  if (RegExAllowedInputExpression.test(value as string)) {
    // This separates out "<+input>.allowedValues(a, b, c)" to ["<+input>", ["a", "b", "c"]]
    const match = (value as string).match(RegExAllowedInputExpression)
    if (match && match?.length > 1) {
      const allowedValues = match[1]
      items.push(...allowedValues.split(',').map(item => ({ label: item, value: item })))
    }
  }
  return items
}

export const shouldRenderRunTimeInputView = (value: any): boolean => {
  if (!value) {
    return false
  }
  if (typeof value === 'object') {
    return Object.keys(value).some(key => typeof value[key] === 'string' && value[key].startsWith(RUNTIME_INPUT_VALUE))
  } else {
    return typeof value === 'string' && value.startsWith(RUNTIME_INPUT_VALUE)
  }
}

export const shouldRenderRunTimeInputViewWithAllowedValues = (
  fieldPath: string,
  template?: Record<string, any>
): boolean => {
  if (!template || !fieldPath) {
    return false
  }
  const allowedValues = get(template, fieldPath, '')
  return shouldRenderRunTimeInputView(allowedValues) && RegExAllowedInputExpression.test(allowedValues)
}

export const getConnectorRefWidth = (viewType: StepViewType): number =>
  Object.entries(ConnectorRefWidth).find(key => key[0] === viewType)?.[1] || ConnectorRefWidth.DefaultView
