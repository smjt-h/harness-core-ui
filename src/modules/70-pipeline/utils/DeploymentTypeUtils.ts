/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import type { ServiceDefinition } from 'services/cd-ng'
import type { StringKeys } from 'framework/strings'

export const deploymentTypeLabel: Record<ServiceDefinition['type'], StringKeys> = {
  Kubernetes: 'kubernetesText',
  NativeHelm: 'pipeline.nativeHelm',
  Ssh: 'SSH',
  ServerlessAwsLambda: 'pipeline.serviceDeploymentTypes.serverlessAwsLambda',
  WinRm: 'pipeline.serviceDeploymentTypes.winrm'
}

export const deploymentTypeIcon: Record<string, IconName> = {
  Kubernetes: 'service-kubernetes',
  NativeHelm: 'service-helm',
  ServerlessAwsLambda: 'service-serverless-aws',
  Ssh: 'secret-ssh',
  WinRm: 'command-winrm'
}
