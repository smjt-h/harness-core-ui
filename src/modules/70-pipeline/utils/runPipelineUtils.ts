/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { cloneDeep, defaultTo } from 'lodash-es'
import type { SelectOption } from '@wings-software/uicore'

import { getStageFromPipeline } from '@pipeline/components/PipelineStudio/PipelineContext/helpers'
import type { AllNGVariables } from '@pipeline/utils/types'
import { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import type { FeaturesProps } from 'framework/featureStore/featureStoreUtil'
import type { UseStringsReturn } from 'framework/strings'
import type { PipelineInfoConfig, StageElementWrapperConfig } from 'services/cd-ng'
import type { InputSetErrorResponse } from 'services/pipeline-ng'

function mergeStage(stage: StageElementWrapperConfig, inputSetPortion: NgPipelineTemplate): StageElementWrapperConfig {
  const stageIdToBeMatched = defaultTo(stage.stage?.identifier, '')
  const matchedStageInInputSet = getStageFromPipeline(stageIdToBeMatched, inputSetPortion.pipeline)

  if (matchedStageInInputSet.stage) {
    let updatedStageVars = []
    if (stage?.stage?.variables && matchedStageInInputSet?.stage?.stage?.variables) {
      updatedStageVars = getMergedVariables(
        stage?.stage?.variables as AllNGVariables[],
        matchedStageInInputSet.stage.stage?.variables as AllNGVariables[]
      )
      matchedStageInInputSet.stage.stage.variables = updatedStageVars
    }
    return matchedStageInInputSet.stage
  }

  return stage
}

interface NgPipelineTemplate {
  pipeline: PipelineInfoConfig
}

export const mergeTemplateWithInputSetData = (
  templatePipeline: NgPipelineTemplate,
  inputSetPortion: NgPipelineTemplate
): NgPipelineTemplate => {
  // Replace all the matching stages in parsedTemplate with the stages received in input set portion
  const mergedStages = templatePipeline.pipeline.stages?.map(stage => {
    if (stage.parallel) {
      /*
      This stage is parallel. Now loop over all the children stages, and check if any of them match in input set portion
      We update all the parallel stages with the ones matching in the input set portion
      and then finally return the new 'updatedParallelStages' object
      */
      const updatedParallelStages = stage.parallel.map(parallelStage => mergeStage(parallelStage, inputSetPortion))
      // Finally setting the updatedParallelStages in the original object, so that the 'mergedStages' will have the updated values
      stage.parallel = updatedParallelStages
      return stage
    }

    /*
    This block will be executed if there are no parallel stages.
    Simply loop over the stages and keep matching and replacing
    */
    return mergeStage(stage, inputSetPortion)
  })

  const toBeUpdated = cloneDeep(templatePipeline)
  toBeUpdated.pipeline.stages = mergedStages
  if (inputSetPortion.pipeline?.properties?.ci) {
    if (!toBeUpdated.pipeline.properties) {
      toBeUpdated.pipeline.properties = {}
    }
    toBeUpdated.pipeline.properties.ci = inputSetPortion.pipeline.properties.ci
  }

  /*
  Below portion adds variables to the pipeline.
  If your input sets has variables, use them.

  Eventually in run pipeline form -
  If input sets are selected, we will supply the variables from 'toBeUpdated' pipleine
  This is why 'toBeUpdated' pipeline should have the variables
  */

  if (inputSetPortion.pipeline.variables) {
    // If we have variables saved in input set, pick them and update

    toBeUpdated.pipeline.variables = getMergedVariables(
      toBeUpdated.pipeline.variables as AllNGVariables[],
      inputSetPortion.pipeline.variables as AllNGVariables[]
    ) // inputSetPortion.pipeline.variables
  }

  return toBeUpdated
}

// Used in Input Set form and save as input set call in run pipeline
export const getFormattedErrors = (apiErrorMap?: { [key: string]: InputSetErrorResponse }): Record<string, any> => {
  const toReturn: Record<string, any> = {}
  if (apiErrorMap) {
    const apiErrorKeys = Object.keys(apiErrorMap)
    apiErrorKeys.forEach(apiErrorKey => {
      const errorsForKey = apiErrorMap[apiErrorKey].errors || []
      if (errorsForKey[0].fieldName) {
        toReturn[errorsForKey[0].fieldName] = `${errorsForKey[0].fieldName}: ${errorsForKey[0].message}`
      }
    })
  }
  return toReturn
}

export const getOverlayErrors = (invalidReferences: Record<string, string>): Record<string, any> => {
  const toReturn: Record<string, any> = {}
  if (invalidReferences) {
    Object.keys(invalidReferences).forEach(invalidReferenceKey => {
      toReturn[invalidReferenceKey] = `${invalidReferenceKey}: ${invalidReferences[invalidReferenceKey]}`
    })
  }

  return toReturn
}

export const getMergedVariables = (
  variables: AllNGVariables[],
  inputSetVariables: AllNGVariables[]
): AllNGVariables[] => {
  // create a map of input set variables values for easier lookup
  // we use "name" of the varibale as the key
  const inputSetVariablesMap: Record<string, AllNGVariables> = inputSetVariables.reduce(
    (acc, curr) => ({ ...acc, [defaultTo(curr.name, '')]: curr }),
    {}
  )

  // loop over existing variables and update their values from input sets
  const finalVariables: AllNGVariables[] = variables.map((variable): AllNGVariables => {
    const { name = '', type } = variable

    // if a variable with same name exists in input set variables
    if (name in inputSetVariablesMap) {
      // copy the variable data
      const newVar: AllNGVariables = { ...inputSetVariablesMap[name] }

      // remove the variable from input set variables
      delete inputSetVariablesMap[name]

      return {
        ...variable,
        // use new value if the type of varibale is same else use the current value
        value: newVar.type === type ? newVar.value : variable.value
      } as AllNGVariables
    }

    // else return original varibale
    return variable
  })

  const remainingVariables = Object.values(inputSetVariablesMap)

  // append the remaining input set variables to existing variables
  return finalVariables.concat(...remainingVariables)
}

export const getRbacButtonModules = (module?: string): string[] => {
  const rbacButtonModules = []
  if (module?.includes('cd')) {
    rbacButtonModules.push('cd')
  }
  if (module?.includes('ci')) {
    rbacButtonModules.push('ci')
  }
  return rbacButtonModules
}
/*
  Get features restriction to pass to 'run/ retry' pipeline button based on the modules the pipeline supports
*/
export const getFeaturePropsForRunPipelineButton = ({
  modules,
  getString
}: {
  modules?: string[]
  getString: UseStringsReturn['getString']
}): FeaturesProps | undefined => {
  if (!modules || !modules?.length) {
    return undefined
  }
  const featureIdentifiers: FeatureIdentifier[] = []
  const additionalFeaturesProps: { warningMessage?: string } = {}
  if (modules.includes('cd')) {
    featureIdentifiers.push(FeatureIdentifier.DEPLOYMENTS_PER_MONTH)
  }
  if (modules.includes('ci')) {
    featureIdentifiers.push(FeatureIdentifier.BUILDS)
    additionalFeaturesProps.warningMessage = getString('pipeline.featureRestriction.unlimitedBuildsRequiredPlan')
  }
  return {
    featuresRequest: {
      featureNames: featureIdentifiers
    },
    ...additionalFeaturesProps
  }
}

export interface SelectedStageData {
  stageIdentifier?: string
  stagesRequired?: string[]
  stageName?: string
  message?: string
}
export interface StageSelectionData {
  selectedStages: SelectedStageData[]
  allStagesSelected: boolean
  selectedStageItems: SelectOption[]
}

export const POLL_INTERVAL = 1 /* sec */ * 1000 /* ms */

export const ALL_STAGE_VALUE = 'all'

export const getAllStageData = (getString: UseStringsReturn['getString']): SelectedStageData => ({
  stageIdentifier: ALL_STAGE_VALUE,
  stagesRequired: [],
  stageName: getString('pipeline.allStages')
})

export const getAllStageItem = (getString: UseStringsReturn['getString']): SelectOption => ({
  label: getString('pipeline.allStages'),
  value: ALL_STAGE_VALUE
})

export function getStageIdentifierFromStageData(selectedStageData: StageSelectionData): string[] {
  return selectedStageData.allStagesSelected
    ? []
    : selectedStageData.selectedStageItems.map(stageData => stageData.value as string)
}
