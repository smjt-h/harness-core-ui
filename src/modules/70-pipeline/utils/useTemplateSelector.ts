/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { parse } from 'yaml'
import { useParams } from 'react-router-dom'
import { isEmpty, merge, set } from 'lodash-es'
import { SelectorData, TemplateDrawerTypes } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import {
  findAllByKey,
  getTemplateTypesByRef,
  usePipelineContext
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import type { TemplateSummaryResponse } from 'services/template-ng'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'

export interface GetTemplateResponse {
  template: TemplateSummaryResponse
  isCopied: boolean
}

export type GetTemplateProps = Omit<SelectorData, 'onSubmit' | 'onCancel'>

interface TemplateActionsReturnType {
  getTemplate: (data: GetTemplateProps) => Promise<GetTemplateResponse>
}

export function useTemplateSelector(): TemplateActionsReturnType {
  const {
    state: { templateTypes, gitDetails },
    updateTemplateView,
    setTemplateTypes
  } = usePipelineContext()
  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()

  const closeTemplateSelector = React.useCallback(() => {
    updateTemplateView({
      isTemplateDrawerOpened: false,
      templateDrawerData: { type: TemplateDrawerTypes.UseTemplate }
    })
  }, [updateTemplateView])

  const updateTemplateTypes = React.useCallback(
    (template: TemplateSummaryResponse, isCopied: boolean) => {
      if (isCopied) {
        const templateRefs = findAllByKey('templateRef', parse(template?.yaml || '')?.template.spec)
        if (!isEmpty(templateRefs)) {
          getTemplateTypesByRef(
            {
              accountIdentifier: accountId,
              orgIdentifier: orgIdentifier,
              projectIdentifier: projectIdentifier,
              templateListType: 'Stable',
              repoIdentifier: gitDetails.repoIdentifier,
              branch: gitDetails.branch,
              getDefaultFromOtherRepo: true
            },
            templateRefs
          ).then(resp => {
            setTemplateTypes(merge(templateTypes, resp))
          })
        }
      } else if (template?.identifier && template?.childType) {
        set(templateTypes, template.identifier, template.childType)
        setTemplateTypes(templateTypes)
      }
    },
    [accountId, orgIdentifier, projectIdentifier, gitDetails, templateTypes, setTemplateTypes]
  )

  const getTemplate = React.useCallback(
    (selectorData: GetTemplateProps): Promise<GetTemplateResponse> => {
      return new Promise((resolve, reject) => {
        updateTemplateView({
          isTemplateDrawerOpened: true,
          templateDrawerData: {
            type: TemplateDrawerTypes.UseTemplate,
            data: {
              selectorData: {
                ...selectorData,
                onSubmit: (template: TemplateSummaryResponse, isCopied: boolean) => {
                  closeTemplateSelector()
                  updateTemplateTypes(template, isCopied)
                  resolve({ template, isCopied })
                },
                onCancel: () => {
                  closeTemplateSelector()
                  reject()
                }
              }
            }
          }
        })
      })
    },
    [updateTemplateView, closeTemplateSelector, updateTemplateTypes]
  )

  return { getTemplate }
}
