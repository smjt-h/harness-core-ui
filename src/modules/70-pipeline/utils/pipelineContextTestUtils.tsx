/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

/**
 * !IMPORTANT
 * Point to Note: Go to the previous versions to see how the functions are actually implemented.
 * Add them back only if they are being unit tested, else leave them out.
 * This is because it unecessarily branches out coverage and we cannot test those that are implemented if they aren't used
 */
import React from 'react'
import { Route } from 'react-router-dom'

import { pick } from 'lodash-es'

import { GitSyncTestWrapper, GitSyncTestWrapperProps } from '@common/utils/gitSyncTestUtils'

import {
  PipelineContext,
  PipelineContextInterface
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'

import pipelineContextProviderProps from './__tests__/mockJson/pipelineContextProvider.json'

export interface PipelineContextWrapperProps extends GitSyncTestWrapperProps {
  pipelineContextValues?: Partial<PipelineContextInterface>
}

jest.mock('services/portal', () => ({
  useGetDelegateSelectorsUpTheHierarchy: jest.fn
}))

jest.mock('services/cd-ng', () => ({
  useGetConnectorListV2: jest.fn(),
  useGetConnectorList: jest.fn(),
  useGetConnector: jest.fn(),
  useGetTestConnectionResult: jest.fn(),
  useGetTestGitRepoConnectionResult: jest.fn(),
  useGetSteps: jest.fn(),
  useGetExecutionStrategyList: jest.fn(),
  useGetExecutionStrategyYaml: jest.fn(),
  useGetServiceList: jest.fn(),
  useGetEnvironmentListForProject: jest.fn(),
  useListGitSync: jest.fn(),
  useGetStepsV2: jest.fn()
}))

jest.mock('services/pipeline-ng', () => ({
  useGetYamlSchema: jest.fn(),
  getPipelinePromise: jest.fn(),
  useGetSteps: jest.fn(),
  useCreateVariablesV2: jest.fn()
}))

export function PipelineContextTestWrapper(
  props: React.PropsWithChildren<PipelineContextWrapperProps>
): React.ReactElement {
  const { defaultAppStoreValues, pipelineContextValues } = props
  return (
    <GitSyncTestWrapper
      {...pick(props, ['path', 'pathParams', 'queryParams'])}
      defaultAppStoreValues={{
        featureFlags: {},
        selectedProject: {
          identifier: 'dummy',
          name: 'dummy',
          modules: ['CD']
        },
        isGitSyncEnabled: true,
        connectivityMode: 'DELEGATE',
        ...defaultAppStoreValues
      }}
    >
      <PipelineContext.Provider
        value={
          {
            fetchPipeline: jest.fn,
            getStageFromPipeline: jest.fn,
            getStagePathFromPipeline: jest.fn,
            setSelection: jest.fn,
            setTemplateTypes: jest.fn,
            updatePipeline: jest.fn,
            updatePipelineView: jest.fn,
            addNewNodeToMap: jest.fn,
            setStepsGraphCanvasState: jest.fn,
            setSelectedStepId: jest.fn,
            setSelectedStageId: jest.fn,
            renderPipelineStage: jest.fn,
            ...pipelineContextProviderProps,
            ...pipelineContextValues,
            state: {
              ...pipelineContextProviderProps.state,
              ...pipelineContextValues?.state
            }
          } as PipelineContextInterface
        }
      >
        <Route exact path={props.path}>
          {props.children}
        </Route>
      </PipelineContext.Provider>
    </GitSyncTestWrapper>
  )
}
