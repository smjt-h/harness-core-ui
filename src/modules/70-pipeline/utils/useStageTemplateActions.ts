/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, isEqual } from 'lodash-es'
import { parse } from 'yaml'
import produce from 'immer'
import { useCallback } from 'react'
import type { StageElementConfig } from 'services/cd-ng'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { useTemplateSelector } from '@pipeline/utils/useTemplateSelector'
import { createTemplate, getStageType } from '@pipeline/utils/templateUtils'
import { getIdentifierFromValue } from '@common/components/EntityReference/EntityReference'

interface TemplateActionsReturnType {
  addOrUpdateTemplate: () => void
  removeTemplate: () => Promise<void>
}

export function useStageTemplateActions(): TemplateActionsReturnType {
  const {
    state: {
      selectionState: { selectedStageId = '' },
      templateTypes
    },
    updateStage,
    getStageFromPipeline
  } = usePipelineContext()
  const { stage } = getStageFromPipeline(selectedStageId)
  const { getTemplate } = useTemplateSelector()

  const addOrUpdateTemplate = useCallback(async () => {
    try {
      const { template, isCopied } = await getTemplate({
        templateType: 'Stage',
        selectedChildType: getStageType(stage?.stage, templateTypes),
        ...(stage?.stage?.template && {
          selectedTemplateRef: getIdentifierFromValue(stage.stage.template.templateRef),
          selectedVersionLabel: stage.stage.template.versionLabel
        })
      })
      const node = stage?.stage
      if (
        !isCopied &&
        isEqual(node?.template?.templateRef, template.identifier) &&
        isEqual(node?.template?.versionLabel, template.versionLabel)
      ) {
        return
      }
      const processNode = isCopied
        ? produce(defaultTo(parse(template?.yaml || '')?.template.spec, {}) as StageElementConfig, draft => {
            draft.name = defaultTo(node?.name, '')
            draft.identifier = defaultTo(node?.identifier, '')
          })
        : createTemplate(node, template)
      await updateStage(processNode)
    } catch (_) {
      // Do nothing.. user cancelled template selection
    }
  }, [stage?.stage, templateTypes, getTemplate, updateStage])

  const removeTemplate = useCallback(async () => {
    const node = stage?.stage
    const processNode = produce({} as StageElementConfig, draft => {
      draft.name = defaultTo(node?.name, '')
      draft.identifier = defaultTo(node?.identifier, '')
      draft.type = getStageType(node, templateTypes)
    })
    await updateStage(processNode)
  }, [stage?.stage, templateTypes, updateStage])

  return { addOrUpdateTemplate, removeTemplate }
}
