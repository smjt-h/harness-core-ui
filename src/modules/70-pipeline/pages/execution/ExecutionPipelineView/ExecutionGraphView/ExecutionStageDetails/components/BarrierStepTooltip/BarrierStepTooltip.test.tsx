/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import BarrierStepTooltip from './BarrierStepTooltip'

describe('BarrierStepTooltip', () => {
  test('if loading', () => {
    const { container } = render(
      <TestWrapper>
        <BarrierStepTooltip loading={true} />
      </TestWrapper>
    )
    const spinner = container.querySelector('.bp3-spinner')
    expect(spinner).toBeTruthy()
  })

  test('basic snapshot after loading', async () => {
    const { container } = render(
      <TestWrapper>
        <BarrierStepTooltip loading={false} startTs={0} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('if startTs at greater than 0', async () => {
    const { queryByText } = render(
      <TestWrapper>
        <BarrierStepTooltip
          loading={false}
          startTs={3}
          data={[
            {
              name: 'Barrier name',
              identifier: 'Barrier identifier'
            }
          ]}
        />
      </TestWrapper>
    )
    await waitFor(() => expect(queryByText('pipeline.barriers.tooltips.barrierWaiting')).toBeTruthy())
  })
})
