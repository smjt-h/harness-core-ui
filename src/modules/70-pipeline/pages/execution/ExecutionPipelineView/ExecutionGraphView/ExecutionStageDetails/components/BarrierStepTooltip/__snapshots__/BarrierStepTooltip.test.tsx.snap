// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarrierStepTooltip basic snapshot after loading 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main StyledProps--border StyledProps--border-top StyledProps--border-color-grey200 StyledProps--padding-right-medium StyledProps--padding-top-small StyledProps--padding-bottom-small StyledProps--padding-left-small"
  >
    <div
      class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-center StyledProps--flex-alignItems-start"
      style="width: 32px;"
    >
      <span
        class="bp3-icon StyledProps--main"
        data-icon="timeout"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 11 11"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 5.5a5.5 5.5 0 11-11 0 5.5 5.5 0 0111 0z"
            fill="#0092E4"
          />
          <path
            clip-rule="evenodd"
            d="M9.746 2.004L5.5 5.5V0c1.71 0 3.237.78 4.246 2.004z"
            fill="#E0E0E0"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main StyledProps--margin-right-medium"
      style="flex: 1;"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-black"
        style="font-size: 12px;"
      />
      <p
        class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey900"
        data-testid="hovercard-service"
        style="font-size: 12px;"
      >
        pipeline.barriers.tooltips.barrierWaiting
      </p>
    </div>
    <div
      class="StyledProps--font StyledProps--main"
    >
      <div
        class="Layout--vertical StyledProps--main StyledProps--background-yellow500 StyledProps--border StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start StyledProps--padding-xsmall"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-weight-bold StyledProps--color StyledProps--color-grey800"
          style="font-size: 10px;"
        >
          0
        </p>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-xsmall StyledProps--color StyledProps--color-grey600"
        >
          min
        </p>
      </div>
    </div>
  </div>
</div>
`;
